---
title: "IHEC RNA-seq and Alternative Splicing"
output: 
  pdf_document: default
  html_notebook: default
---

Libraries and Helper-Functions:

```{r}
library(data.table)
# data.table::setDTthreads(20)
library(ggplot2)
library(ggfortify)
library(ggrepel)

colVar <- function(x, na.rm = TRUE) {
  colMeans(x * x, na.rm = na.rm) - (colMeans(x, na.rm = na.rm)) ^ 2
}

pca_plots <-
  function(matrix,
           metadata,
           title = '',
           color_by = 'annotation',
           shape_by = 'project',
           label = 'annotation',
           na.omit = FALSE) {
    if (na.omit){
      matrix <- matrix[, colSums(is.na(matrix)) == 0, drop=FALSE]
    } else {
      # set NA to 0
      matrix[is.na(matrix)] <- 0
    }
    
    # filter out cols without variance
    matrix <- matrix[, colVar(matrix) != 0, drop = FALSE]
    
    # perform PCA
    pca <- prcomp(matrix, center = TRUE, scale = TRUE)
    
    # compute explained variance of PCs
    var_explained_tpm = 100 * pca$sdev ^ 2 / sum(pca$sdev ^ 2)
    
    return(
      list(
        pca = pca,
        dimred = ggplot2::autoplot(
          pca,
          data = metadata,
          colour = color_by,
          shape = shape_by
        ) + labs(title = paste('PCA:', title)) + geom_text_repel(
          aes(label = get(label)),
          min.segment.length = 0,
          max.overlaps = 5
        ),
        scree = qplot(seq.int(var_explained_tpm), var_explained_tpm) +
          geom_line() +
          xlab("Principal Component") +
          ylab("Percent Variance Explained") +
          ggtitle(paste("Variance explained:", title))
      )
    )
  }
```

Load Metadata:

```{r}
# load metadata
metadata <- data.table::fread('IHEC_metadata_harmonization.v0.7.csv')
metadata[, EpiRR_no_version := data.table::tstrsplit(EpiRR, '.', fixed = TRUE)[1]]
metadata[, annotation := cell_type]
metadata[biomaterial_type == 'primary tissue', annotation := tissue_type]
metadata[biomaterial_type == 'cell line', annotation := line]
metadata[annotation == "CD34-positive, CD38-positive common myeloid progenitor OR CD34-positive, CD38-positive common lymphoid progenitor", annotation := "CD34-positive, CD38-positive common progenitor"]
```

Load Isoform Quantifications:

```{r}
rna_metadata <- fread(file.path(data_dir, 'ihec_metadata_rna.csv'))
rna_metadata[, EpiRR_no_version := tstrsplit(epirr_id, '.', fixed = TRUE)[1]]
multiple_RNAs <- rna_metadata[, if(.N == 2) .(uuid, rna_seq_type), by=EpiRR_no_version]

isoform_files <- list.files(rna_data_dir, pattern = '\\.isoforms\\.results$', full.names = T)
isoform_files <- isoform_files[!grepl(multiple_RNAs[rna_seq_type == 'total-RNA-Seq', paste(uuid, collapse = '|')], isoform_files)]
keep_cols <- c('transcript_id', 'gene_id', 'TPM', 'FPKM')
isoform_quants <- data.table::rbindlist(sapply(isoform_files, data.table::fread, stringsAsFactor=TRUE, simplify = FALSE, select = keep_cols), idcol = 'EpiRR')
file2id <- gsub("^.*?(IHECRE[0-9]{8}).*$", "\\1", isoform_files)
names(file2id) <- isoform_files
isoform_quants[, EpiRR_no_version := as.factor(file2id[isoform_quants[, EpiRR]])]

gene_quants <- isoform_quants[, .(gene_tpm = sum(TPM), transcripts = transcript_id), by = c('EpiRR_no_version', 'gene_id')]
gene_expr_file <- 'suppa_analysis/gene_expressions.tsv'
if (!file.exists(gene_expr_file)) {
  fwrite(gene_quants, gene_expr_file)
}

# bring tpms to wide format
tpm_dt <- data.table::dcast(isoform_quants, EpiRR_no_version ~ transcript_id, value.var = 'TPM')


# remove id column
tpm_matrix <- as.matrix(tpm_dt[, -'EpiRR_no_version'])

# write files for SUPPA2
rownames(tpm_matrix) <- tpm_dt[, EpiRR_no_version]
suppa_expr_filename <- 'suppa_analysis/tpm_expressions.tsv'
if (!file.exists(suppa_expr_filename)) {
  write.table(t(tpm_matrix), file=suppa_expr_filename, sep="\t", quote = FALSE)
}

# filter metadata
metadata <- metadata[EpiRR_no_version %in% rownames(tpm_matrix)]
# remove the two outliers
metadata_filtered <- metadata[cell_type != 'sperm']
```

# RUN 02-AS-Analysis here


## PCA on TPM

```{r, fig.height=7, fig.width=12}
pca_plots(tpm_matrix, metadata, 'TPM w/ outliers')
```

```{r, fig.height=7, fig.width=12}

pca_plots(tpm_matrix[metadata_filtered[, EpiRR_no_version]], metadata_filtered, 'TPM w/o outliers')
```

## UMAP on TPM

```{r, fig.height=7, fig.width=12}
library(umap)
umap_tpm <- umap::umap(tpm_matrix[metadata_filtered[, EpiRR_no_version]])

df <- data.frame(x = umap_tpm$layout[,1],
                 y = umap_tpm$layout[,2],
                 annotation = metadata_filtered$annotation,
                 project = metadata_filtered$project)

ggplot(df, aes(x, y, color = annotation, shape = project, label=annotation)) +
  geom_point() +
  geom_text_repel( min.segment.length = 0, max.overlaps = 5) +
  labs(x = "UMAP1", y = "UMAP2", title = 'UMAP: TPM w/o outliers')
```

## PCA on Isoform Usage

```{r, fig.height=7, fig.width=12}
isoform_usages <- t(as.matrix(read.table('suppa_analysis/all_isoform.psi', header=TRUE, sep="\t")))

pca_plots(isoform_usages, metadata, 'IU w/ outliers')
```

```{r, fig.height=7, fig.width=12}
pca_plots(isoform_usages[metadata_filtered[, EpiRR_no_version],], metadata_filtered, 'IU w/o outliers')
```

## PCA on AS events (PSI values)

```{r}
psi_files <- list.files('suppa_analysis/events', pattern = '\\.psi$', full.names = TRUE)
psi_table <- do.call(cbind, lapply(psi_files, function(f) t(as.matrix(read.table(f, header=TRUE, sep="\t")))))

psi_files_tpm <- list.files('suppa_analysis/events/TPM1', pattern = '\\.psi$', full.names = TRUE)
psi_table_tpm <- do.call(cbind, lapply(psi_files_tpm, function(f) t(as.matrix(read.table(f, header=TRUE, sep="\t")))))
```

```{r, fig.height=7, fig.width=12}
na_count <- apply(psi_table, 2, function(y) sum(is.na(y)))
na_count_tpm <- apply(psi_table_tpm, 2, function(y) sum(is.na(y)))

dt <- data.table(na_count = c(na_count, na_count_tpm),
                 event = c(names(na_count),names(na_count_tpm)),
                 filter = c(rep(FALSE, length(na_count)), rep(TRUE, length(na_count))))
dt[, event_type := gsub('^.*;(.*?):.*', '\\1', event)]

ggplot(dt, aes(x = na_count, fill = filter)) + 
  geom_histogram(position='dodge', binwidth = 2) +
  facet_wrap(. ~ event_type, scales= 'free')

ggplot(dt, aes(x = na_count, fill = filter)) + 
  geom_density(alpha= .5, binwidth = 2) +
  facet_wrap(. ~ event_type, scales= 'free')
```

```{r}
psi_dt_tpm <- as.data.table(psi_table_tpm)
psi_dt_tpm[, EpiRR_no_version := rownames(psi_table_tpm)]
psi_dt_tpm <- melt(psi_dt_tpm, id.vars = 'EpiRR_no_version', variable.name = 'event', value.name = 'PSI')
print(sprintf('for all %i events, there are %i with no NAs (%2.2f%%)', psi_dt_tpm[, uniqueN(event)], psi_dt_tpm[, .(no_na = all(!is.na(PSI))), by=event][, sum(no_na)], 100 * psi_dt_tpm[, .(no_na = all(!is.na(PSI))), by=event][, sum(no_na)] / psi_dt_tpm[, uniqueN(event)] ))
psi_dt_tpm[event %in% psi_dt_tpm[is.na(PSI), unique(event)], as.list(summary(PSI)), by = event]
```

```{r, fig.height=7, fig.width=12}
psi_pca_tpm <- pca_plots(psi_table_tpm[metadata_filtered[, EpiRR_no_version],], metadata_filtered, 'PSI w/o outliers w/ TPM filter setting NAs to 0')
print(psi_pca_tpm$dimred)
psi_pca_tpm_omit <- pca_plots(psi_table_tpm[metadata_filtered[, EpiRR_no_version],], metadata_filtered, 'PSI w/o outliers w/ TPM filter ommitting events with NAs', na.omit = TRUE)
print(psi_pca_tpm_omit$dimred)
```

```{r, fig.height=7, fig.width=12}
pca_plots(psi_table, metadata, 'PSI w/ outliers')
```

```{r, fig.height=7, fig.width=12}
out <- pca_plots(psi_table[metadata_filtered[, EpiRR_no_version],], metadata_filtered, 'PSI w/o outliers')
out$dimred
```

```{r, fig.height=7, fig.width=12}
pca_plots(psi_table_tpm[metadata_filtered[, EpiRR_no_version],], metadata_filtered, 'PSI w/o outliers w/ TPM filter')
```

```{r, fig.height=7, fig.width=12, eval=TRUE}
library(pheatmap)
pheatmap(psi_table_filtered)
```

```{r}
psi_dt <- as.data.table(psi_table)
psi_dt[, EpiRR_no_version := rownames(psi_table)]
psi_dt <- merge(metadata[, .(EpiRR_no_version, annotation)], psi_dt, by ='EpiRR_no_version')
psi_dt_melt <- melt(psi_dt, id.vars = c('EpiRR_no_version', 'annotation'), variable = 'event', value = 'PSI')
```
