---
title: "preliminary-results"
output: pdf_document
---

```{r}
# first get all rna, histone and wgbs data to match the samples
rna_samples <- fread(file.path(data_dir, 'ihec_metadata_rna.csv'))
rna_samples[, epirr_id_wo_version := tstrsplit(epirr_id, '.', fixed = TRUE)[1]]

histone_samples <- fread(file.path(data_dir, 'ihec_metadata.csv'))
histone_samples[, epirr_id_wo_version := tstrsplit(epirr_id, '.', fixed = TRUE)[1]]

histone_samples_available <- sort(unique(gsub("^.*?(IHECRE[0-9]{8}).*$", "\\1", list.files(chip_data_dir, '\\.fc\\.signal\\.bigwig$'))))

if (length(setdiff(histone_samples[, epirr_id_wo_version], histone_samples_available)) > 0) {
  print(paste('in metadata file but no signal file:', paste(histone_samples[!epirr_id_wo_version %in% histone_samples_available, epirr_id_wo_version], collapse = ', ')))
  print(paste('signal file but not in metadata file:', paste(histone_samples_available[!histone_samples_available %in% histone_samples[, epirr_id_wo_version]], collapse = ', ')))
}

rna_samples_w_chip <- rna_samples[epirr_id_wo_version %in% histone_samples_available]
# there are some sample where there is mRNA and total-RNA-Seq. in that case we take mRNA-Seq
rna_samples_w_chip <- rna_samples_w_chip[, if(.N > 1){.SD[rna_seq_type != 'total-RNA-Seq']}else{.SD}, by = epirr_id_wo_version]

wgbs_samples <- fread(file.path(data_dir, 'ihec_metadata_wgbs.csv'))
wgbs_samples[, epirr_id_wo_version := tstrsplit(epirr_id, '.', fixed = TRUE)[1]]


wgbs_samples_available <- sort(unique(gsub("^.*?(IHECRE[0-9]{8}).*$", "\\1", list.files(wgbs_data_dir, '\\.gembs_(pos|neg)\\.bw$'))))

if (length(setdiff(wgbs_samples[, epirr_id_wo_version], wgbs_samples_available)) > 0) {
  print(paste('in metadata file but no signal file:', paste(wgbs_samples[!epirr_id_wo_version %in% wgbs_samples_available, epirr_id_wo_version], collapse = ', ')))
  print(paste('signal file but not in metadata file:', paste(wgbs_samples_available[!wgbs_samples_available %in% wgbs_samples[, epirr_id_wo_version]], collapse = ', ')))
}

rna_samples_w_wgbs <- rna_samples[epirr_id_wo_version %in% wgbs_samples_available]
# there are some sample where there is mRNA and total-RNA-Seq. in that case we take mRNA-Seq
rna_samples_w_wgbs <- rna_samples_w_wgbs[, if(.N > 1){.SD[rna_seq_type != 'total-RNA-Seq']}else{.SD}, by = epirr_id_wo_version]

```


```{r}
event_res <- sapply(as_events, function(event) {
	# load psi values
  event_dt <- fread(paste0('suppa_analysis/events/TPM1/event_', event, '.psi'))
  setnames(event_dt, 'V1', 'event')
  # get all the samples that are also in the metadata and filter
  samples_to_consider <- intersect(names(event_dt), rna_samples[, epirr_id_wo_version])
  event_dt <- event_dt[, c('event', ..samples_to_consider)]
  # remove all the events where there is only one valid value. They would be discarded later anyway
  event_dt <- event_dt[apply(event_dt[, ..samples_to_consider], 1, function(r) sum(is.na(r)) < (length(r) - 1)), ]
  # add ID
  event_dt[, ID:=seq.int(.N)]
  # prepare the alternative/upstream/downstream region
  if (event == 'SE') {
    event_dt[, c('event_name', 'seqnames', 'e1s2', 'e2s3', 'strand') := tstrsplit(event, ':', fixed=TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(event_name, ';', fixed=T)]
    event_dt[, c('e1', 's2', 'e2', 's3') := c(tstrsplit(e1s2, '-', fixed = TRUE, type.convert = TRUE), tstrsplit(e2s3, '-', fixed = TRUE, type.convert = TRUE))]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = e1, end = s3), strand = strand)]
    upstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', e1, e2), end = ifelse(strand == '+', s2, s3)), strand = strand)]
    downstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', e2, e1), end = ifelse(strand == '+', s3, s2)), strand = strand)]
    event_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = s2, end = e2), strand = strand)]
  } else if (event == 'RI') {
    event_dt[, c('event_name', 'seqnames', 's1', 'e1s2', 'e2', 'strand') := tstrsplit(event, ':', fixed=TRUE, type.convert = TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(event_name, ';', fixed=T)]
    event_dt[, c('e1', 's2') := tstrsplit(e1s2, '-', fixed = TRUE, type.convert = TRUE)]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = s1, end = e2), strand = strand)]
    upstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', s1, s2), end = ifelse(strand == '+', e2, e2)), strand = strand)]
    downstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', s2, s1), end = ifelse(strand == '+', e2, e1)), strand = strand)]
    event_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = e1, end = s2), strand = strand)]
  } else {
    stop('unknown event')
  }
  
  # filter out events from multiple genes or overlapping events
  self_hits <- findOverlaps(all_gr)
  non_overlapping <- self_hits@from[which(!(duplicated(self_hits@from) | duplicated(self_hits@from, fromLast = TRUE)))]
  single_genes_only <- event_dt[-event_dt[, grep('_and_', gene_id, fixed = TRUE)], ID]
  single_genes_non_overlapping <- intersect(non_overlapping, single_genes_only)
  list(event_dt = event_dt[, c('ID', 'gene_id', 'seqnames', 'strand', ..samples_to_consider)], event_gr = event_gr, upstream_gr = upstream_gr, downstream_gr = downstream_gr, single_genes_non_overlapping = single_genes_non_overlapping)
}, simplify = FALSE)
```

```{r, fig.width = 12, fig.height = 12}
library(pheatmap)

samples_with_epi <- intersect(rna_samples_w_chip[, epirr_id_wo_version], rna_samples_w_wgbs[, epirr_id_wo_version])
var_events <- sapply(names(event_res), function(event) {
  this_event <- event_res[[event]]
  event_mat <- as.matrix(this_event[['event_dt']][, ..samples_with_epi])
  event_mat[is.nan(event_mat)] <- NA
  events_to_consider <- intersect(this_event[['single_genes_non_overlapping']], which(apply(event_mat, 1, function(r) sum(is.na(r)) < (length(r) - 1))))
  
  # check event psi variation
  event_psi_sd <- apply(event_mat, 1, sd, na.rm = TRUE)
  hist(event_psi_sd, 30, main = event)
  quants <- seq(0, .9, .1)
  var_events <- sapply(quants, function(quant) {
    threshold <- quantile(event_psi_sd, quant, na.rm=TRUE)
    var_events <- intersect(events_to_consider, which(event_psi_sd >= threshold))
  })
  names(var_events) <- quants
  
  rmv <- apply(event_mat, 1, function(r) sum(is.na(r)) > ncol(event_mat)/2)
  print(pheatmap(event_mat[intersect(var_events[['0.5']], which(!rmv)), ], show_colnames = FALSE, main = event))
  var_events
}, simplify = FALSE)
```

```{r}
# merge event_dts and gr objects. adjust IDs
max_ID <- 0L
event_dt <- data.table()
event_gr <- GRanges()
upstream_gr <- GRanges()
downstream_gr <- GRanges()
single_genes_non_overlapping <- integer()
for (event_name in names(event_res)) {
  this_dt <- event_res[[event_name]]$event_dt
  this_max_ID <- this_dt[, max(ID)]
  var_events[[event_name]] <- sapply(var_events[[event_name]], function(ids) ids + max_ID, simplify = FALSE)
  single_genes_non_overlapping <- c(single_genes_non_overlapping, event_res[[event_name]]$single_genes_non_overlapping + max_ID)
  event_gr <- c(event_gr, event_res[[event_name]]$event_gr)
  upstream_gr <- c(upstream_gr, event_res[[event_name]]$upstream_gr)
  downstream_gr <- c(downstream_gr, event_res[[event_name]]$downstream_gr)
  this_dt[, ID := ID + max_ID]
  this_dt[, event_name := event_name]
  event_dt <- rbind(event_dt, this_dt)
  max_ID <- max_ID + this_max_ID
}
stopifnot(event_dt[, max(ID)] == nrow(event_dt))
```

```{r}
if (file.exists('aggregated_regions.rds')) {
  aggregated_data <- readRDS('aggregated_regions.rds')
} else {
  source('aggregating_helper.R')
  samples_to_consider <- names(event_dt)[startsWith(names(event_dt), "IHECRE") ]
  aggregated_data <- aggregate_multiple_samples(samples_to_consider, event_dt, event_gr, upstream_gr, downstream_gr)
  names(aggregated_data) <- samples_to_consider
  saveRDS(aggregated_data, 'aggregated_regions.rds')
}

# at some point rename event_name and other_region with ifelse(event_name == 'SE', 'intron', 'exon')
aggregated_data <- aggregated_data[samples_with_epi]
```


```{r, fig.width = 12, fig.height = 12}
library(corrplot)
ihec <- 'IHECRE00001878'
this_event_dt <- aggregated_data[[ihec]]
names(this_event_dt) <- gsub(' ', '_', names(this_event_dt), fixed = TRUE)
aggregation <- 'max'
cor_method <- 'pearson'
keep_cols <- function(dt, aggregation_method) {!grepl(paste(names(aggregation_functions)[names(aggregation_functions) != aggregation_method], collapse = '|'), names(dt))}
max_event_cols <- keep_cols(this_event_dt, aggregation)
event_mat <- as.matrix(this_event_dt[, ..max_event_cols][, -'ID'])
colnames(event_mat) <- gsub('_max', '', colnames(event_mat), fixed = TRUE)
event_mat[, 'gene_expression'] <- log2(event_mat[, 'gene_expression'] + 1)
for (event_type in names(var_events)) {
  var_events_type <- var_events[[event_type]]
  for (var_threshold in names(var_events_type))
  corrplot(cor(event_mat[var_events_type[[var_threshold]],], use = 'na', method = cor_method), order = 'hclust', title = paste0('\n\n', ihec, '\n', event_type, ' variace_threshold: ', var_threshold))
}
```

```{r}
metadata <- data.table::fread('IHEC_metadata_harmonization.v0.7.csv')
metadata[, EpiRR_no_version := data.table::tstrsplit(EpiRR, '.', fixed = TRUE)[1]]
metadata[, annotation := cell_type]
metadata[biomaterial_type == 'primary tissue', annotation := tissue_type]
metadata[biomaterial_type == 'cell line', annotation := line]
metadata[annotation == "CD34-positive, CD38-positive common myeloid progenitor OR CD34-positive, CD38-positive common lymphoid progenitor", annotation := "CD34-positive, CD38-positive common progenitor"]
```

```{r, fig.width=10, fig.height=8}
library(pheatmap)
library(ggplot2)
for (entry in aggregated_data) {
  if (any(duplicated(sub('^reH', 'H', names(entry))))) {
    drop_cols <- names(entry)[startsWith(names(entry), 're') & (duplicated(sub('^reH', 'H', names(entry))) | duplicated(sub('^reH', 'H', names(entry)), last = TRUE))]
    entry[, (drop_cols):=NULL]
  } else {
    setnames(entry, names(entry), sub('^reH', 'H', names(entry)))
  }
  setnames(entry, names(entry), gsub(' ', '_', names(entry), fixed = TRUE))
}
aggregated_dt <- rbindlist(aggregated_data, idcol = 'IHEC', fill = TRUE)
aggregated_dt[, gene_expression:=log2(gene_expression + 1)]
aggregated_dt[is.nan(PSI), PSI := NA]
aggregated_dt[event_dt, on = .(ID), event := event_name]

for (event_name in event_dt[, unique(event_name)]) 
print(ggplot(aggregated_dt[event == event_name & IHEC %in% aggregated_dt[, sample(unique(IHEC), size = 16)]], aes(x = PSI)) + geom_histogram(color = 'white', binwidth = .05) + theme_bw() + facet_wrap(. ~ IHEC) + labs(title = event_name))

unlisted_vars <- unlist(unname(var_events), recursive = FALSE)
cor_dt <- rbindlist(sapply(unique(names(unlisted_vars)), function(var){
  var_ids <- unlist(unlisted_vars[names(unlisted_vars) == var], use.names = FALSE)
  aggregated_dt[ID %in% var_ids, lapply(.SD[, -c('ID', 'PSI')], function(y) suppressWarnings(cor(y, PSI, use = 'na', method = cor_method))), by = .(event, IHEC)]
}, simplify = FALSE), idcol = 'variance quartile')

max_cor_cols <- keep_cols(cor_dt, aggregation)
cor_dt_intermediate <- cor_dt[, ..max_cor_cols]
colnames(cor_dt_intermediate) <- gsub('_max', '', colnames(cor_dt_intermediate), fixed = TRUE)
melt_cor_dt <- melt(cor_dt_intermediate, id.vars = c('IHEC', 'variance quartile', 'event'), variable.name = 'feature', value.name = 'correlation w/ PSI')
melt_cor_dt[event == 'SE', feature := gsub('other_region', 'intron', feature, fixed=TRUE)]
melt_cor_dt[event == 'RI', feature := gsub('other_region', 'exon', feature, fixed=TRUE)]
melt_cor_dt[event == 'SE', feature := gsub('event_name', 'SE', feature, fixed=TRUE)]
melt_cor_dt[event == 'RI', feature := gsub('event_name', 'RI', feature, fixed=TRUE)]
melt_cor_dt[, median_cor:=median(`correlation w/ PSI`, na.rm = TRUE), by = c('feature', 'event', 'variance quartile')]
melt_cor_dt[, relevant:=any(abs(median_cor > .05)), by = c('event', 'feature', 'variance quartile')]

subset_cor_dt <- melt_cor_dt[`variance quartile` %in% c('0', '0.9')]
subset_cor_dt[, relevant := any(relevant), by = .(feature, event)]

breaks <- seq(-.2, .2, by = .01)
print(ggplot(subset_cor_dt[relevant == TRUE], aes(x = `correlation w/ PSI`, y = reorder(feature, -`correlation w/ PSI`, FUN = median, na.rm = TRUE), color = `variance quartile`)) + theme_bw() + geom_boxplot(aes(fill = median_cor)) + scale_fill_gradientn(colours = colorRampPalette(c("red", "white", "blue"))(length(breaks)), breaks = seq(-.2, .2, by = .1), limits = c(-.2, .2)) + scale_colour_manual(values = c('dark grey', 'black')) + ylab('feature') + facet_grid(event ~ ., scales="free_y"))
```


```{r, fig.width = 18, fig.height = 8, eval=FALSE}
library(pheatmap)
library(ggplot2)
for (entry in se_raw) {
  if (any(duplicated(sub('^reH', 'H', names(entry))))) {
    drop_cols <- names(entry)[startsWith(names(entry), 're') & (duplicated(sub('^reH', 'H', names(entry))) | duplicated(sub('^reH', 'H', names(entry)), last = TRUE))]
    entry[, (drop_cols):=NULL]
  } else {
    setnames(entry, names(entry), sub('^reH', 'H', names(entry)))
  }
  setnames(entry, names(entry), gsub(' ', '_', names(entry), fixed = TRUE))
}
se_dt <- rbindlist(se_raw, idcol = 'IHEC', fill = TRUE)
se_dt[, gene_expression:=log2(gene_expression + 1)]
se_dt[is.nan(PSI), PSI := NA]
ggplot(se_dt[IHEC %in% se_dt[, sample(unique(IHEC), size = 16)]], aes(x = PSI)) + geom_histogram(color = 'white', binwidth = .05) + theme_bw() + facet_wrap(. ~ IHEC)

for (var_threshold in names(var_events)) {
  cor_dt <- se_dt[ID %in% var_events[[var_threshold]], lapply(.SD[, -c('ID', 'PSI')], function(y) suppressWarnings(cor(y, PSI, use = 'na', method = cor_method))), by = 'IHEC']
  max_cor_cols <- keep_cols(cor_dt, aggregation)
  
  cor_dt_intermediate <- cor_dt[, ..max_cor_cols]
  colnames(cor_dt_intermediate) <- gsub('_max', '', colnames(cor_dt_intermediate), fixed = TRUE)
  cor_mat <- as.matrix(cor_dt_intermediate[, -'IHEC'])
  rownames(cor_mat) <- cor_dt_intermediate[, IHEC]
  breaks <- seq(-.2, .2, by = .01)
  annotation_df <- metadata[EpiRR_no_version %in% samples_with_epi, data.frame(annotation = annotation, row.names = EpiRR_no_version)]
  pheatmap(t(cor_mat[samples_with_epi, ])[, rownames(annotation_df)[order(annotation_df$annotation)]], color = colorRampPalette(c("red", "white", "blue"))(length(breaks)), breaks = breaks, annotation_col = annotation_df, cluster_cols = FALSE, main = paste('PSI Variance Quantile:', var_threshold))
  melt_cor_dt <- melt(cor_dt_intermediate, id.vars = 'IHEC', variable.name = 'feature', value.name = 'correlation w/ PSI')
  melt_cor_dt[, median_cor:=median(`correlation w/ PSI`, na.rm = TRUE), by = feature]
  print(ggplot(melt_cor_dt, aes(y = `correlation w/ PSI`, x = reorder(feature, -`correlation w/ PSI`, FUN = median, na.rm = TRUE))) + theme_bw() + geom_boxplot(aes(fill = median_cor)) + theme(axis.text.x = element_text(angle = 90, hjust=1)) + scale_fill_gradientn(colours = colorRampPalette(c("red", "white", "blue"))(length(breaks)), breaks = seq(-.2, .2, by = .1), limits = c(-.2, .2)) + xlab('feature') + ggtitle(paste('PSI Variance Quantile:', var_threshold)))
}
```

```{r}
se_max_cols <- keep_cols(se_dt, aggregation)
feature_data <- se_dt[, ..se_max_cols][, -c('IHEC', 'ID')]
response <- 'PSI'
explanatory <- names(feature_data)[names(feature_data) != response]
feature_data[, (names(feature_data)) := lapply(names(feature_data), function(x) scale(get(x)))]
linear_model <- lm(formula(paste(response, '~', paste(explanatory, collapse = ' + '))), feature_data)
summary(linear_model)
```

