---
title: "preliminary-results"
output: pdf_document
---


```{r}
psi_input_dir <- "suppa_analysis/events"
header <- readLines('suppa_analysis/tpm_expressions.tsv', n=1)
sample_cols <- strsplit(header, '\t', TRUE)[[1]]
# all_event_files <- list.files(psi_input_dir, pattern = paste0('[^', paste(as_events, collapse = '|'), ']\\.psi$'), full.names = TRUE)
# all_event_dt <- rbindlist(sapply(all_event_files, fread, simplify=FALSE))
# setnames(other_event_dt, 'V1', 'event')
event_res <- sapply(as_events, function(event_type) {
	# load psi values
  event_dt <- fread(paste0(psi_input_dir, '/event_', event_type, '.psi'))
  setnames(event_dt, 'V1', 'event_id')
  # # get all the samples that are also in the metadata and filter
  # event_dt <- event_dt[, c('event', ..samples_with_epi)]
  # # remove all the events where there is only one valid value. They would be discarded later anyway
  # event_dt <- event_dt[apply(event_dt[, ..sample_cols], 1, function(r) sum(is.na(r)) < (length(r) - 1)), ]
  
  # prepare the alternative/upstream/downstream region
  if (event_type == 'SE') {
    event_dt[, c('Event Type', 'seqnames', 'e1s2', 'e2s3', 'strand') := tstrsplit(event_id, ':', fixed=TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(`Event Type`, ';', fixed=T)]
    event_dt[, c('e1', 's2', 'e2', 's3') := c(tstrsplit(e1s2, '-', fixed = TRUE, type.convert = TRUE), tstrsplit(e2s3, '-', fixed = TRUE, type.convert = TRUE))]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = e1, end = s3), strand = strand)]
    upstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', e1, e2), end = ifelse(strand == '+', s2, s3)), strand = strand)]
    downstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', e2, e1), end = ifelse(strand == '+', s3, s2)), strand = strand)]
    event_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = s2, end = e2), strand = strand)]
  } else if (event_type == 'RI') {
    event_dt[, c('Event Type', 'seqnames', 's1', 'e1s2', 'e2', 'strand') := tstrsplit(event_id, ':', fixed=TRUE, type.convert = TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(`Event Type`, ';', fixed=T)]
    event_dt[, c('e1', 's2') := tstrsplit(e1s2, '-', fixed = TRUE, type.convert = TRUE)]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = s1, end = e2), strand = strand)]
    upstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', s1, s2), end = ifelse(strand == '+', e2, e2)), strand = strand)]
    downstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', s2, s1), end = ifelse(strand == '+', e2, e1)), strand = strand)]
    event_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = e1, end = s2), strand = strand)]
  } else if (event_type %in% c('AL', 'AF')){
    event_dt[, c('Event Type', 'seqnames', 's1/e1s2', 'e1s3/e2', 's2/e1s3', 'e2s3/e3', 'strand') := tstrsplit(event_id, ':', fixed=TRUE, type.convert = TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(`Event Type`, ';', fixed=T)]
    event_dt[ifelse(event_type=='AF', strand == '+', strand == '-'), start_all:=as.integer(`s1/e1s2`)]
    event_dt[ifelse(event_type=='AF', strand == '-', strand == '+'), start_all:=tstrsplit(`s1/e1s2`, '-', fixed=TRUE, type.convert = TRUE)[[1]]]
    
    event_dt[ifelse(event_type=='AF', strand == '-', strand == '+'), end_all:=as.integer(`e2s3/e3`)]
    event_dt[ifelse(event_type=='AF', strand == '+', strand == '-'), end_all:=tstrsplit(`e2s3/e3`, '-', fixed=TRUE, type.convert = TRUE)[[2]]]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = start_all, end = end_all), strand = strand)]
    upstream_gr <- NULL
    downstream_gr <- NULL
    event_gr <- NULL
  } else if (event_type %in% c('A3', 'A5')){
    event_dt[, c('Event Type', 'seqnames', 'e1s2/e2s3', 'e1s3', 'strand') := tstrsplit(event_id, ':', fixed=TRUE, type.convert = TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(`Event Type`, ';', fixed=T)]
    event_dt[, c('e1/e2', 's2/s3') := tstrsplit(`e1s2/e2s3`, '-', fixed = TRUE, type.convert = TRUE)]
    event_dt[, c('e1', 's3') := tstrsplit(e1s3, '-', fixed = TRUE, type.convert = TRUE)]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = `e1/e2`, end = s3), strand = strand)]
    upstream_gr <- NULL
    downstream_gr <- NULL
    event_gr <- NULL
  } else if (event_type == 'MX'){
    event_dt[, c('Event Type', 'seqnames', 'e1s2', 'e2s4', 'e1s3', 'e3s4', 'strand') := tstrsplit(event_id, ':', fixed=TRUE, type.convert = TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(`Event Type`, ';', fixed=T)]
    event_dt[, start_all:=tstrsplit(e1s2, '-', fixed=TRUE, type.convert = TRUE)[[1]]]
    event_dt[, end_all:=tstrsplit(e3s4, '-', fixed=TRUE, type.convert = TRUE)[[2]]]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = start_all, end = end_all), strand = strand)]
    upstream_gr <- NULL
    downstream_gr <- NULL
    event_gr <- NULL
  } else stop('unknown event')
  
  list(event_dt = event_dt[, c('event_id', 'gene_id', 'seqnames', 'strand', ..sample_cols)], event_gr = event_gr, upstream_gr = upstream_gr, downstream_gr = downstream_gr, all_gr=all_gr)
}, simplify = FALSE)
```

```{r, fig.width=7, fig.height=4}
library(UpSetR)
library(grid)

all_gr <- Reduce(c, lapply(event_res, function(event){event$all_gr}))

# merge event_dts and gr objects. adjust IDs
# max_ID <- 0L
event_dt <- data.table()
event_gr <- GRanges()
upstream_gr <- GRanges()
downstream_gr <- GRanges()
for (this_event in to_analyze) {
  this_dt <- copy(event_res[[this_event]]$event_dt)
  this_event_gr <- event_res[[this_event]]$event_gr
  single_genes_only <- this_dt[!grepl('_', gene_id, fixed = TRUE), which = TRUE]
  not_gonosomes <- this_dt[!seqnames %in% c('chrY', 'chrX'), which = TRUE]
  # var_events[[this_event]] <- sapply(var_events[[this_event]], function(ids) ids + max_ID, simplify = FALSE)
  
  other_event_overlaps <- findOverlaps(this_event_gr, all_gr)
  non_overlapping_other <- other_event_overlaps@from[which(!(duplicated(other_event_overlaps@from) | duplicated(other_event_overlaps@from, fromLast = TRUE)))]
  self_event_overlaps <- findOverlaps(this_event_gr)
  non_overlapping_self <- self_event_overlaps@from[which(!(duplicated(self_event_overlaps@from) | duplicated(self_event_overlaps@from, fromLast = TRUE)))]  
  pdf(file=file.path(plot_dir, paste0(this_event, '_event_intersections.pdf')), width = 7, height = 5, onefile = FALSE) # or other device
  upset <- upset(fromList(list('all'=this_dt[, seq.int(.N)],'no overlap other'=non_overlapping_other, 'no overlap self'=non_overlapping_self, 'single gene'=single_genes_only, 'not gonosomes' = not_gonosomes)), sets.bar.color = "#56B4E9", order.by = "freq")
  print(upset)
  grid.text(this_event, x = 0.65, y=0.95, gp=gpar(fontsize=20)) 
  dev.off()
  print(upset)
  grid.text(this_event, x = 0.65, y=0.95, gp=gpar(fontsize=20)) 
  # keep_rows <- intersect(single_genes_only, non_overlapping_other)
  # add ID
  # this_dt <- this_dt[keep_rows]
  # this_dt[, ID:=seq.int(.N)]
  
  # this_max_ID <- this_dt[, max(ID)]
  
  event_gr <- c(event_gr, this_event_gr)#[keep_rows])
  upstream_gr <- c(upstream_gr, event_res[[this_event]]$upstream_gr)#[keep_rows])
  downstream_gr <- c(downstream_gr, event_res[[this_event]]$downstream_gr)#[keep_rows])
  # this_dt[, ID := ID + max_ID]
  this_dt[, `Event Type` := this_event]
  event_dt <- rbind(event_dt, this_dt)
  # max_ID <- max_ID + this_max_ID
}
event_dt[, ID:=seq.int(.N)]
event_dt[, `Event Type`:=factor(`Event Type`, levels=c('SE', 'RI'))]

other_event_overlaps <- findOverlaps(event_gr, all_gr)
non_overlapping_other <- other_event_overlaps@from[which(!(duplicated(other_event_overlaps@from) | duplicated(other_event_overlaps@from, fromLast = TRUE)))]

single_genes_only <- event_dt[!grepl('_', gene_id, fixed = TRUE), which = TRUE]
not_gonosomes <- event_dt[!seqnames %in% c('chrY', 'chrX'), which = TRUE]

keep_rows <- intersect(intersect(single_genes_only, non_overlapping_other), not_gonosomes)

stopifnot(identical(event_dt[, ID], seq(nrow(event_dt))))# && max_ID == event_dt[, max(ID)])
cols_to_factor <- c('event_id', 'gene_id', 'seqnames', 'strand', 'Event Type')
event_dt[,(cols_to_factor) := lapply(.SD, as.factor), .SDcols = cols_to_factor]
```


```{r, fig.width=7, fig.height=7}
saveRDS(event_res, 'event_res.rds')
rm(event_res)
```

```{r, fig.width = 12, fig.height = 12}
var_events <- sapply(to_analyze, function(event) {
  this_event_dt <- event_dt[`Event Type` == event]
  event_mat <- as.matrix(this_event_dt[ID %in% keep_rows, ..sample_cols])
  rownames(event_mat) <- this_event_dt[ID %in% keep_rows, ID]
  event_mat[is.nan(event_mat)] <- NA
  na_count <- apply(event_mat, 1, function(r) sum(is.na(r))) # < (length(r) - 1)))
  na_percentage <- na_count / length(sample_cols)
  
  # check event psi variation
  event_psi_sd <- apply(event_mat, 1, sd, na.rm = TRUE)
  event_psi_mean <- apply(event_mat, 1, mean, na.rm = TRUE)
  # hist(event_psi_sd, 30, main = event)
  print(ggplot(data.table(sd = event_psi_sd, na_percentage=na_percentage), aes(x = na_percentage, y = sd)) + geom_point() + labs(title=event, x=paste0('na percentage (overall ', length(sample_cols), ')')) + theme_bw())
  sd_mean_dt <- data.table(sd = event_psi_sd, event_psi_mean=event_psi_mean)
  quants <- seq(0, .9, .1)
  var_events <- sapply(quants, function(quant) {
    threshold <- quantile(event_psi_sd, quant, na.rm=TRUE)
    idx <- which(event_psi_sd >= threshold)
    stopifnot(identical(names(idx), rownames(event_mat)[idx]))
    sd_mean_dt[idx, percentage:=as.character(quant)]
    names(idx)
  })
  names(var_events) <- quants
  print(ggplot(sd_mean_dt, aes(x = event_psi_mean, y = sd, color = percentage)) + geom_hline(yintercept = quantile(event_psi_sd, quants, na.rm=TRUE, names=FALSE)) + geom_point() + labs(title=event, x='mean psi') + theme_bw())
  
  
  print(pheatmap(event_mat, show_colnames = FALSE, show_rownames = FALSE, main = event, cluster_rows = FALSE, cluster_cols = FALSE))
  rmv <- apply(event_mat, 1, function(r) sum(is.na(r)) > length(r)/2)
  print(pheatmap(event_mat[intersect(var_events[['0.5']], names(rmv)[!rmv]), ], show_colnames = FALSE, main = paste(event, '0.5 var threshold')))
  var_events
}, simplify = FALSE)
```

```{r}
library(ggplot2)
ggplot(melt(event_dt[ID %in% keep_rows, sapply(sample_cols, function(x) sum(is.nan(get(x)))/.N, simplify = FALSE), by = `Event Type`], measure.vars = sample_cols, variable.name = 'IHEC', value.name = 'na_proportion'), aes(x = `Event Type`, y = na_proportion)) + geom_violin() + labs(y = 'Proportion of NAs in PSI Values per Epigenome', x ='Event Name') + geom_dotplot(binaxis='y', stackdir='center', dotsize=.5) + geom_boxplot(width=0.1) + stat_summary(fun=mean, geom="point", shape=23, size=2) + scale_y_continuous(breaks = seq(0, 1, .1))+ theme_bw()
ggsave(filename = file.path(plot_dir, 'sample_nas.pdf'))
# melt(event_dt, id.vars = c('ID', 'gene_id', 'seqnames', 'strand', 'Event Type'), variable.name = 'IHEC', value.name = 'PSI')
ggplot(melt(event_dt[ID %in% keep_rows], id.vars = c('ID', 'gene_id', 'seqnames', 'strand', 'Event Type'), measure.vars = sample_cols, variable.name = 'IHEC', value.name = 'PSI')[, .(na_proportion=sum(is.nan(PSI))/.N), by = .(ID, `Event Type`)], aes(x = `Event Type`, y = na_proportion)) + geom_violin() + labs(y = 'Proportion of NAs in PSI Values per Event', x ='Event Type')  + geom_dotplot(binaxis='y', stackdir='center', dotsize=.1, binwidth = .01) + geom_boxplot(width=0.1) + stat_summary(fun=mean, geom="point", shape=23, size=2) + scale_y_continuous(breaks = seq(0, 1, .1))+ theme_bw()
ggsave(filename = file.path(plot_dir, 'event_nas.pdf'))
ggplot(melt(event_dt[ID %in% keep_rows], id.vars = c('ID', 'gene_id', 'seqnames', 'strand', 'Event Type'), measure.vars = sample_cols, variable.name = 'IHEC', value.name = 'PSI')[, .(na_proportion=sum(is.nan(PSI))/(.N)), by = .(ID, `Event Type`)], aes(x = na_proportion)) + facet_wrap(. ~ `Event Type`, scales = 'free') + stat_ecdf() + labs(title = 'na proportion by events')
```

```{r}
annotation <- rtracklayer::import('suppa_analysis/gencode.v29.annotation.gtf')
used_genes <- annotation[annotation$gene_id %in% event_dt[, gene_id] & annotation$type == 'gene', ]
event_dt[data.table(gene_id = used_genes$gene_id, gene_start = start(used_genes), gene_end = end(used_genes)), on = 'gene_id', c('gene_start', 'gene_end') := .(gene_start, gene_end)]

event_dt[strand == '-', distance_TSS:=log2(gene_end - end(event_gr[ID]))]
event_dt[strand != '-', distance_TSS:=log2(start(event_gr[ID]) - gene_start)]

event_dt[strand == '-', distance_TES:=log2(start(event_gr[ID]) - gene_start)]
event_dt[strand != '-', distance_TES:=log2(gene_end - end(event_gr[ID]))]

# add stats about enhancers
# signal_psi_dt[, count_enhancer := as.integer(table(c(signal_psi_dt[, ID], from(enhancer_hits)))) - 1L]
    
# signal_psi_dt[data.table(ID = from(distances_enhancers), distance = mcols(distances_enhancers)$distance), on = .(ID), distance_enhancer := distance]

# load gene quantifications
gene_quants <- readRDS('suppa_analysis/gene_expressions.rds')

# separator_gene_ids <- '_and_'
# signal_psi_dt <- event_dt[, .(ID, PSI = get(ihec), gene_id)]
# signal_psi_dt <- signal_psi_dt[, .(gene_id = unlist(tstrsplit(gene_id, separator_gene_ids, fixed=TRUE))), by=.(ID, PSI)]
# signal_psi_dt[unique(gene_quants[EpiRR_no_version == ihec, .(gene_id, gene_tpm)]), on = .(gene_id), gene_expression := gene_tpm]
# signal_psi_dt <- signal_psi_dt[, .(gene_id=paste(gene_id, collapse=separator_gene_ids), gene_expression=sum(gene_expression, na.rm = TRUE)), by=.(ID, PSI)]

    
# add region_lengths
event_dt[, paste('width', 'upstream_other_region', sep = ';') := log2(width(upstream_gr))]
event_dt[, paste('width', 'event_name', sep = ';') := log2(width(event_gr))]
event_dt[, paste('width', 'downstream_other_region', sep = ';') := log2(width(downstream_gr))]

# add distance to TSS and TES here
# signal_psi_dt[, distance_TSS := distance_TSS]
# signal_psi_dt[width(promoter_gr) == 1, distance_TSS := NA] # no distance for multi-gene-exons
# signal_psi_dt[, distance_TES := distance_TES]
# signal_psi_dt[width(tes_gr) == 1, distance_TES := NA] # no distance for multi-gene-exons
```

```{r}
metadata <- data.table::fread('IHEC_metadata_harmonization.v1.0.extended.csv')
metadata[, EpiRR_no_version := data.table::tstrsplit(EpiRR, '.', fixed = TRUE)[1]]
metadata[, annotation := harmonized_sample_ontology_intermediate]
# metadata[biomaterial_type == 'primary tissue', annotation := tissue_type]
# metadata[biomaterial_type == 'cell line', annotation := line]
# metadata[annotation == "CD34-positive, CD38-positive common myeloid progenitor OR CD34-positive, CD38-positive common lymphoid progenitor", annotation := "CD34-positive, CD38-positive common progenitor"]
```


```{bash}
mkdir -p data
cd data 
wget https://downloads.wenglab.org/Registry-V3/GRCh38-cCREs.bed
```

```{r}
# aggregated_file_path <- file.path(psi_input_dir, 'aggregated_reference.rds')
save.image('aggregating.rda')
```

```{r}
# if (!file.exists(aggregated_file_path)) {
#   source('04-preliminary-results-aggregation.R') # or run with sbatch 04-preliminary-results-aggregation.sh
# }
# 
# aggregated_data <- readRDS(aggregated_file_path)

# at some point rename `Event Type` and other_region with ifelse(`Event Type` == 'SE', 'intron', 'exon')

# aggregated_data <- aggregated_data[sample_cols]
```
