---
title: "Post-aggregation analysis"
output: pdf_document
---

```{r, fig.width=16, fig.height=10}
load('aggregating.rda')
file_table <- fread('file_table.csv.gz')
file_table[, filename:=file.path('sample_dts', basename(filename))]
file_table[, filename:=paste0(filename, '.tab.gz')]
file_table <- file_table[experiment_type != "DNAm"]
setkey(file_table, 'filename')

wgbs_agg <- fread('sample_dts/WGBS_agg.csv.gz')
wgbs_agg <- wgbs_agg[!startsWith(name, 'chromhmm')]
wgbs_agg[, experiment_type:='DNAm']

histone_dt <- rbindlist(
  pbmcapply::pbmcmapply(
    FUN = function(file) {
      dt <-
        fread(file)
      names(dt) <-
        c('name', 'size', 'covered', 'sum', 'mean0', 'mean', 'min', 'max')
      dt[!startsWith(name, 'chromhmm')]
    },
    file = file_table[, filename],
    USE.NAMES = TRUE,
    SIMPLIFY = FALSE
  ),
  idcol = 'filename'
)
histone_dt[, ihec:=file_table[filename, epirr_id]]
histone_dt[, experiment_type:=file_table[filename, experiment_type]]
setnames(histone_dt, old = 'mean', new= 'score')

common_ihec <- intersect(histone_dt[, unique(ihec)], wgbs_agg[, unique(ihec)])

common_cols <- intersect(names(histone_dt), names(wgbs_agg))
```


```{r, fig.width=16, fig.height=10}
aggregated_dt <- rbindlist(list(histone_dt[ihec %in% common_ihec, ..common_cols], wgbs_agg[ihec %in% common_ihec, ..common_cols]))
aggregated_dt[, c("position", "ID"):=tstrsplit(name, "_", fixed=TRUE)]
aggregated_dt[is.na(ID), c("position", "ID"):=list('event_name', position)]
aggregated_dt[position=="up", position:="upstream_other_region"]
aggregated_dt[position=="down", position:="downstream_other_region"]
aggregated_dt[, ihec:=as.factor(ihec)]
aggregated_dt[, ID:=as.integer(ID)]
aggregated_dt <- dcast(aggregated_dt, ihec + ID ~ experiment_type + position, value.var = "score", sep = ";")
setnames(aggregated_dt, old = 'ihec', new = 'IHEC')
aggregated_dt[, IHEC:=gsub(pattern = "\\..*", "", IHEC)]
```


```{r, fig.width=16, fig.height=10, eval=FALSE}
agg_data_files <- list.files(sample_dt_dir, pattern = 'merge_dt.csv.gz', full.names = TRUE)
agg_data_files <- agg_data_files[!startsWith(basename(agg_data_files), 'old')]
histone_marks <- c('H3K4me1', 'H3K27me3', 'H3K36me3', 'H3K27ac', 'H3K4me3', 'H3K9me3')
cCRE_regions <- c('CTCF-only', 'dELS', 'pELS', 'PLS', 'DNase-H3K4me3')
cCRE_regions <- c(cCRE_regions, paste(cCRE_regions, 'CTCF-bound', sep = ','))
genomic_regions <- c('upstream_other_region', 'event_name', 'downstream_other_region')
all_regions <- c(cCRE_regions, genomic_regions)
histone_cols <- paste(CJ(histone_marks, all_regions)[, paste(histone_marks, all_regions, sep=';')], 'max', sep = ';')
wgbs_cols <- paste('wgbs', all_regions, 'mean', sep = ';')
cols_to_read <- c('ID', histone_cols, wgbs_cols)
file2IHEC <- tstrsplit(basename(agg_data_files), '-', fixed=TRUE, keep = 1)[[1]]
names(file2IHEC) <- agg_data_files
agg_file_list <- suppressWarnings(pbmcapply::pbmclapply(agg_data_files, fread, select = cols_to_read))
names(agg_file_list) <- agg_data_files
aggregated_dt <- rbindlist(agg_file_list, idcol = 'IHEC', fill = TRUE)
aggregated_dt[, IHEC:= as.factor(file2IHEC[IHEC])]
```


```{r, fig.width=16, fig.height=10}
# for (j in which(!(names(aggregated_dt) %in% c('IHEC', 'ID') | startsWith(names(aggregated_dt), "DNAm"))))
#   set(aggregated_dt,NULL,j,log2(min(aggregated_dt[[j]][aggregated_dt[[j]] != 0], na.rm = TRUE) + aggregated_dt[[j]]))
# for (entry in aggregated_data) {
#   if (any(duplicated(sub('^reH', 'H', names(entry))))) {
#     drop_cols <- names(entry)[startsWith(names(entry), 're') & (duplicated(sub('^reH', 'H', names(entry))) | duplicated(sub('^reH', 'H', names(entry)), last = TRUE))]
#     entry[, (drop_cols):=NULL]
#   }
#   setnames(entry, names(entry), sub('^reH', 'H', names(entry)))
#   rare_antibodies <- names(entry)[grep('H3K9/14ac|H2A\\.Zac', names(entry))]
#   entry[, (rare_antibodies):=NULL]
#   # setnames(entry, names(entry), gsub(' ', '_', names(entry), fixed = TRUE))
# }
# aggregated_dt <- rbindlist(aggregated_data, idcol = 'IHEC', fill = TRUE)
# rm(aggregated_data)
# aggregated_dt[, IHEC:=as.factor(IHEC)]
fwrite(aggregated_dt, 'aggregated_dt.csv.gz')
# aggregated_dt <- aggregated_dt[ID %in% keep_rows]
# aggregated_dt <- aggregated_dt[!is.nan(PSI)]
# rm(agg_file_list)
gc()

# if (!file.exists('aggregated_dt.rds'))
# saveRDS(aggregated_dt, 'aggregated_dt.rds')
```

```{r, height=10, fig.width=8, eval=FALSE}
cCREs <- data.table::fread('data/GRCh38-cCREs.bed')
names(cCREs) <- c('seqnames', 'start', 'end', 'some_id', 'accession', 'cCRE_type')
cCREs[, ID:=.I]
cCRE_gr <- cCREs[, GRanges(seqnames = seqnames, IRanges(start = start, end = end))]
cCRE_hits <- findOverlaps(event_gr, cCRE_gr, maxgap = vicinity, ignore.strand=TRUE)
hits_used <- cCRE_hits[from(cCRE_hits) %in% keep_rows]
cCREs[, 'neighborhood':='In SCREEN']
cCREs[unique(from(hits_used)), 'neighborhood':='In Event Neighborhood']
ggplot(cCREs, aes(x = reorder(cCRE_type, cCRE_type, function(x) -length(x)))) + geom_bar() + theme_bw() + theme(axis.text.x = element_text(angle = 45, hjust=1)) + labs(x = 'cCRE Type') + facet_wrap( ~ `neighborhood`, nrow = 2, scales = 'free_y')
ggsave(filename = file.path(plot_dir, 'cCRE_distr.pdf'), width = 7, height = 5)
# our_cCRE <- ggplot(cCREs[unique(from(hits_used))], aes(x = reorder(cCRE_type, cCRE_type, function(x) -length(x)))) + geom_bar() + theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust=1)) + labs(title='Number of Occurences per cCRE Type in Event Neighborhood', x = 'cCRE Type')


n_cCREs_long <- data.table(ID=from(hits_used), cCRE_type=cCREs[from(hits_used), cCRE_type])[, .(n_cCRE=.N), by=.(ID, cCRE_type)]
n_cCREs_long[event_dt, on=.(ID), `Event Type`:=`Event Type`]
n_cCREs_long[, `Event Type`:=factor(`Event Type`, levels=c('SE', 'RI'))]
n_cCREs_long[,cCRE_type:=sub(pattern = ',', replacement = '\n', cCRE_type, fixed = TRUE)]
ggplot(n_cCREs_long, aes(x = n_cCRE, fill = cCRE_type)) + geom_histogram(binwidth = 1, color = 'white') + facet_wrap(~ `Event Type`, scales='free') + theme_bw() + labs(x='#cCRE in Event Neighborhood', y='Count') + scale_fill_discrete(name = "cCRE Type")
ggsave(filename = file.path(plot_dir, 'cCRE_counts.pdf'), width = 8, height = 4)
n_cCREs_long[, n_cCRE:=log2(n_cCRE)]
n_cCREs <- dcast(n_cCREs_long, ID ~ cCRE_type, value.var = 'n_cCRE', fill = 0L)
n_cCRE_cols <- names(n_cCREs)[names(n_cCREs) != 'ID']
new_n_cCRE_cols <- paste('#cCRE', n_cCRE_cols, sep =';')
# event_dt[n_cCREs, on=.(ID), (new_n_cCRE_cols):=mget(n_cCRE_cols)]wgbs event_dt[ID
# %in% keep_rows & is.na(n_cCRE), n_cCRE:=0]
```


```{r, fig.width=16, fig.height=10}
melted_psi <- melt(event_dt, measure.vars = sample_cols, id.vars = 'ID', variable.name = 'IHEC', value.name = 'PSI', variable.factor = FALSE)
aggregated_dt[melted_psi[ID %in% keep_rows & !is.na(PSI)], on=.(ID, IHEC), PSI:=PSI]
aggregated_dt <- aggregated_dt[!is.na(PSI)]

cols_to_add <- c('gene_id', 'distance_TSS', 'distance_TES', 'width;upstream_other_region', 'width;event_name', 'width;downstream_other_region', 'Event Type')
aggregated_dt[event_dt, on=.(ID), (cols_to_add):=mget(cols_to_add)]
aggregated_dt[, `Event Type`:=factor(`Event Type`, levels=c('SE', 'RI'))]
aggregated_dt[gene_quants, on=c(gene_id='gene_id', IHEC='EpiRR_no_version'), gene_expression:=gene_tpm]
aggregated_dt[, gene_expression:=log2(gene_expression)]
aggregated_dt[, gene_id:=NULL]
```


```{r, fig.width=16, fig.height=10}
percentiles <- c(1/3, .5, 2/3)
sample_quartiles <- aggregated_dt[, split(quantile(PSI, probs = percentiles, names= FALSE), as.character(round(percentiles, 2))), by = .(`Event Type`, IHEC)]
melted_sample_quartiles <- melt(sample_quartiles, id.vars = c('Event Type', 'IHEC'), variable.name='Percentage', value.name = 'quantile')
print(ggplot(melted_sample_quartiles, aes(x = quantile, fill = Percentage)) + geom_histogram(binwidth = 0.01) + theme_bw() + theme(legend.position = 'bottom') + facet_wrap(~`Event Type`, scales='free_y') + labs(x='PSI Quantile', y='Count'))
ggsave(filename = file.path(plot_dir, 'percentage_dist.pdf'), width = 6, height = 3)
event_quartiles <- aggregated_dt[, .(Mean=mean(PSI), `SD`=sd(PSI), `NA Percentage`= 1- length(IHEC)/aggregated_dt[, uniqueN(IHEC)]), by = .(`Event Type`, ID)]
print(ggplot(event_quartiles, aes(x = Mean, y = SD, color = `Event Type`)) + geom_point() + geom_hline(data=event_quartiles[, .(median_sd=median(SD, na.rm = TRUE)), by = `Event Type`], aes(yintercept = median_sd, color = `Event Type`)) + theme_bw())
ggsave(filename = file.path(plot_dir, 'PSI_mean_sd.pdf'), width = 8, height = 5)

print(ggplot(event_quartiles, aes(x = `NA Percentage`, y = SD, color = `Event Type`)) + geom_point() + geom_smooth(aes(color = `Event Type`), method='lm') + ggpubr::stat_cor(aes(color = `Event Type`)) + theme_bw())
ggsave(filename = file.path(plot_dir, 'PSI_sd_na.pdf'), width = 8, height = 4)

for (this_event in to_analyze) {
  print(ggplot(aggregated_dt[`Event Type` == this_event & IHEC %in% aggregated_dt[, sample(unique(IHEC), size = 16)]], aes(x = PSI)) + geom_histogram(color = 'white', binwidth = .05) + theme_bw() + facet_wrap(. ~ IHEC) + labs(title = this_event, y='Count'))
  
  print(ggplot(aggregated_dt[ID %in% aggregated_dt[`Event Type` == this_event, sample(unique(ID), size = 16)]], aes(x = PSI)) + geom_histogram(color = 'white', binwidth = .05) + stat_bin(aes(y=..count.., label=ifelse(..count..==0,"",..count..)), geom="text", hjust=-.5, angle = 90, binwidth = .05) + theme_bw() + facet_wrap(. ~ ID) + labs(title = this_event, y='Count'))
}
```

```{r, fig.width=16, fig.height=10}
safe_colorblind_palette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", 
                             "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")
id_vars <- c('IHEC', 'ID', 'Event Type')
features <- setdiff(names(aggregated_dt), id_vars)
marks <- unique(tstrsplit(features, ';', fixed=TRUE, keep = 1)[[1]])
print(ggplot(melt(aggregated_dt, id.vars = c('IHEC', 'ID', 'Event Type'), measure.vars = features, variable.name = 'feature', value.name = 'value')[, .(sd=sd(value, na.rm=TRUE)), by=.(`Event Type`, feature)], aes(y = reorder(feature, sd), x = sd)) + geom_point() + facet_wrap(.~`Event Type`, ncol=2) + labs(x = 'SD', y = 'Variable') + theme_bw())# + theme(axis.text.x = element_text(angle=, hjust=1)))
ggsave(filename = file.path(plot_dir, 'feature_dists',  'sds.pdf'), width = 5, height = 16)

# ggplot(agg_melt, aes(x = value, color = feature)) + geom_density() + facet_wrap(. ~ `Event Type`, nrow=2) + geom_text_repel(
#     aes(label = sd(value)))

for (mark in marks) {
  agg_melt <-
    melt(
      aggregated_dt,
      id.vars = c('IHEC', 'ID', 'Event Type'),
      measure.vars = features[startsWith(features, mark)]
    )
  if (any(agg_melt[, grepl(';', variable, fixed = TRUE)])) {
    agg_melt[, c('mark', 'region') := tstrsplit(variable, ';', fixed = TRUE, keep = 2)]
    agg_melt[region == 'event_name', region := 'Alternative Region']
    agg_melt[region == 'downstream_other_region', region := 'Downstream Region']
    agg_melt[region == 'upstream_other_region', region := 'Upstream Region']
    agg_melt[, region := factor(
      region,
      levels =
        c(
          unique(region)[startsWith(unique(region), 'Upstream')],
          unique(region)[startsWith(unique(region), 'Alternative')],
          unique(region)[startsWith(unique(region), 'Downstream')]
          # unique(region)[startsWith(unique(region), 'PLS')],
          # unique(region)[startsWith(unique(region), 'DNase-H3K4me3')],
          # unique(region)[startsWith(unique(region), 'pELS')],
          # unique(region)[startsWith(unique(region), 'dELS')],
          # unique(region)[startsWith(unique(region), 'CTCF-only')]
        )
    )]
  } else {
    agg_melt[, region := 'Gene-Wide']
    agg_melt[, mark := variable]
  }
  # max_dt <- agg_melt[, {
  #   dens <- density(value, na.rm=TRUE)
  #   .(xMax=(dens)$x[which.max(density(value, na.rm=TRUE)$y)], yMax=max((dens)$y, na.rm=TRUE))
  #   }, by=.(`Event Type`, region)]
  this_colors <- safe_colorblind_palette[seq_len(agg_melt[, uniqueN(region)])]
  p <-
    ggplot(agg_melt, aes(x = value, color = region, label = region)) +
    geom_density() + 
    facet_wrap(. ~ `Event Type`, nrow = 2, scales = 'free_y') + 
    theme_bw() + 
    labs(color = 'Region') + 
    scale_color_manual(values = this_colors) + 
    guides(color = guide_legend(override.aes = list(fill = this_colors))) # + geom_text_repel(data=max_dt, aes(xMax, yMax, color=region, label=region), min.segment.length = 0, force=75)
    if (mark == 'DNAm') {
      p <- p + labs(title = 'DNAm', x = 'log2(Mean Methylation)')
    } else if (mark == 'distance_TSS') {
      p <- p + labs(title = 'Distance to TSS', x = 'log2(bp)')
    } else if (mark == 'distance_TES') {
      p <- p + labs(title = 'Distance to TES', x = 'log2(bp)')
    } else if (mark == 'width') {
      p <- p + labs(title = 'Width', x = 'log2(bp)')
    } else if (mark == 'gene_expression') {
      p <- p + labs(title = 'Gene Expression', x = 'log2(TPM)')
    } else if (mark == 'PSI') {
      p <- p + labs(x = 'PSI')
    } else {
      p <- p + labs(title = mark, x = '-log10pval') + scale_x_log10()
    }
  print(p)
  ggsave(
    filename = file.path(plot_dir, 'feature_dists',  paste0(mark, '.pdf')),
    width = 8,
    height = 4
  )
}
```

```{r, fig.width=16, fig.height=10}
# if (!file.exists('aggregated_dt_filtered.rds'))
# saveRDS(aggregated_dt, 'aggregated_dt_filtered.rds')
# fwrite(event_dt[ID %in% keep_rows], "events.csv.gz")
fwrite(aggregated_dt, 'aggregated_dt_filtered.csv.gz')
```
