---
title: "IHEC RNA-seq and Alternative Splicing"
output: 
  pdf_document: default
  html_notebook: default
header-includes:
- \usepackage{xcolor}
---

```{r}
library(data.table)
# data.table::setDTthreads(20)
library(ggplot2)
library(ggfortify)
library(ggrepel)

colVar <- function(x, na.rm=TRUE) {
  colMeans(x*x, na.rm=na.rm) - (colMeans(x, na.rm=na.rm))^2
}

pca_plots <- function(matrix, metadata, title=NULL, color_by='annotation', shape_by='project', label='annotation') {
  # set NA to 0 
  matrix[is.na(matrix)] <- 0
  
  # filter out cols without variance
  matrix <- matrix[, colVar(matrix) != 0, drop = FALSE]
  
  # perform PCA
  pca <- prcomp(matrix, center = TRUE, scale = TRUE)
  
  # compute explained variance of PCs
  var_explained_tpm = 100 * pca$sdev^2 / sum(pca$sdev^2)
  
  return(list(dimred=ggplot2::autoplot(pca, data=metadata, colour=color_by, shape = shape_by) + labs(title = paste('PCA:', title)) + geom_text_repel(aes(label=get(label)), min.segment.length = 0, max.overlaps = 5),
              scree=qplot(seq.int(var_explained_tpm), var_explained_tpm) + 
                geom_line() + 
                xlab("Principal Component") + 
                ylab("Percent Variance Explained") +
                ggtitle(paste("Variance explained:", title))))
}
```

```{r}
isoform_files <- list.files('/nfs/data/IHEC/RNAseq', full.names = T)
keep_cols <- c('transcript_id', 'gene_id', 'TPM', 'FPKM')
isoform_quants <- data.table::rbindlist(sapply(isoform_files, data.table::fread, stringsAsFactor=TRUE, simplify = FALSE, select = keep_cols), idcol = 'sample')

# bring tpms to wide format
tpm_wide <- data.table::dcast(isoform_quants, sample ~ transcript_id, value.var = 'TPM')
tpm_wide[, sample:=as.factor(gsub("^.*?(IHECRE[0-9]{8}).*$", "\\1", sample))]

# load metadata
metadata <- data.table::fread('IHEC_metadata_harmonization.v0.5.csv')
metadata[, EpiRR_merging := data.table::tstrsplit(EpiRR, '.', fixed = TRUE)[1]]
tpm_wide <- data.table::merge.data.table(tpm_wide, metadata, all.x = TRUE, by.x = 'sample', by.y='EpiRR_merging')
tpm_wide[, annotation := cell_type]
tpm_wide[biomaterial_type == 'primary tissue', annotation := tissue_type]
tpm_wide[biomaterial_type == 'cell line', annotation := line]

# keep only tpm columns
transcript_cols <- names(tpm_wide)[startsWith(names(tpm_wide), 'ENST')]
tpm_matrix <- as.matrix(tpm_wide[, ..transcript_cols])

# write files for SUPPA2
rownames(tpm_matrix) <- tpm_wide[, sample]
suppa_expr_filename <- 'suppa_analysis/tpm_expressions.tsv'
if (!file.exists(suppa_expr_filename)) {
  write.table(t(tpm_matrix), file=suppa_expr_filename, sep="\t", quote = FALSE)
}
```

## PCA on TPM
```{r, fig.height=7, fig.width=12}
# filter out transcripts where all samples are smaller than 1
# tpm_matrix <- tpm_matrix[, colSums(tpm_matrix < 1) < nrow(tpm_matrix), drop = FALSE]

pca_plots(tpm_matrix, tpm_wide, 'TPM w/ outliers')
```

```{r, fig.height=7, fig.width=12}
# remove two outliers
tpm_wide_filtered <- tpm_wide[cell_type != 'sperm cell']
tpm_matrix_filtered <- as.matrix(tpm_wide_filtered[, ..transcript_cols])

# filter again in case outliers dominated the expression of certain transcripts
# tpm_matrix_filtered <- tpm_matrix_filtered[, colSums(tpm_matrix_filtered < 1) < nrow(tpm_matrix_filtered), drop = FALSE]

pca_plots(tpm_matrix_filtered, tpm_wide_filtered, 'TPM w/o outliers')
```

## UMAP on TPM
```{r, fig.height=7, fig.width=12}
library(umap)
umap_tpm <- umap::umap(tpm_matrix_filtered)

df <- data.frame(x = umap_tpm$layout[,1],
                 y = umap_tpm$layout[,2],
                 annotation = tpm_wide_filtered$annotation,
                 project = tpm_wide_filtered$project)

ggplot(df, aes(x, y, color = annotation, shape = project, label=annotation)) +
  geom_point() +
  geom_text_repel( min.segment.length = 0, max.overlaps = 5) +
  labs(x = "UMAP1", y = "UMAP2", title = 'UMAP: TPM w/o outliers')
```


## PCA on Isoform Usage

```{r, fig.height=7, fig.width=12}
isoform_usages <- t(as.matrix(read.table('suppa_analysis/all_isoform.psi', header=TRUE, sep="\t")))

pca_plots(isoform_usages, tpm_wide, 'IU w/ outliers')
```

```{r, fig.height=7, fig.width=12}
isoform_usages_filtered <- isoform_usages[tpm_wide_filtered[, sample],]

pca_plots(isoform_usages_filtered, tpm_wide_filtered, 'IU w/o outliers')
```

## PCA on AS events (PSI values)

```{r, fig.height=7, fig.width=12}
psi_files <- list.files('suppa_analysis/events', pattern = '\\.psi$', full.names = TRUE)
psi_table <- do.call(cbind, lapply(psi_files, function(f) t(as.matrix(read.table(f, header=TRUE, sep="\t")))))

pca_plots(psi_table, tpm_wide, 'PSI w/ outliers')
```

```{r, fig.height=7, fig.width=12}
psi_table_filtered <- psi_table[tpm_wide_filtered[, sample],]

pca_plots(psi_table_filtered, tpm_wide_filtered, 'PSI w/o outliers')
```